### PR TITLE
Prevented add-ons from being enabled if NVDA has been started with --disable-addons flag

### LINKED ISSUE #9473: AddonsDialog: some add-ons are marked as Enabled when NVDA is launched with --disable-addons flag
### Steps to reproduce:

1. Launch NVDA.
2. Make sure you have at least one add-on installed.
3. Close NVDA.
4. Launch NVDA from the command line with the `--disable-addons` flag.
5. Open the add-on manager dialog.

### Actual behavior:

Add-ons that are enabled on your profile will appear as enabled on the dialog.

### Expected behavior:

Since NVDA has been told to start with all add-ons disabled, and since they are effectively not running, they shouldn't appear as enabled.

I found this when looking at #5785.

Since a label indicates that you must restart NVDA to re-enable your add-ons, and because the `--disable-addons` will likely be used by people who know what they're doing, I am wondering if we shouldn't disable this dialog and its menu entry all together.

#### NVDA installed/portable/running from source:

NVDA Installed.

#### NVDA version:

NVDA 2019.1.

#### Windows version:

Windows 10 Education version 1809.

#### Does the issue still occur after restarting your PC?

Yes.

#### Have you tried any other versions of NVDA?

No.

### PR DESCRIPTION
**Pull request by DataTriny**
### Link to issue number:

Fixes #9473.

### Summary of the issue:

When NVDA is launched with the `--disable-addons` flag, add-ons that were previously enabled will still appear as such in the add-ons dialog. Furthermore, it is possible to disable and enable them.

### Description of how this pull request fixes the issue:

- An add-on is now considered disabled if the `disableAddons` global flag is on, or the add-on is in the disabled add-ons list.
- The enable/disable button in the add-ons dialog has been disabled if the `disableAddons` global flag is on.

### Testing performed:

- Installed several add-ons, made sure that the actual behavior was untouched.
- Restarted NVDA with `--disable-addons` and ensured that all add-ons appeared disabled and that it was not possible anymore to trigger the enable button.

### Known issues with pull request:

None

### Change log entry:

Section: Bug fixes

- When NVDA is launched with the `--disable-addons` flag, all add-ons now appears disabled.

### CODE REVIEW

**Comment by DrSooom** [comment]
@DataTriny: Is it still possible to uninstall an add-on when starting NVDA with the --disable-addons flag? I guess it should be still possible due to very broken add-ons, which maybe can't be uninstalled while they are enabled.

**Comment by DrSooom** [comment]
One more question, because I haven't tested this: What about the install button? Should we disable it too? Then we would have to disable the installation method via the Windows Explorer (by opening an .nvda-addon file) too. That would be another PR.

File: source/addonHandler/__init__.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # -*- coding: UTF-8 -*-
 #addonHandler.py
 #A part of NonVisual Desktop Access (NVDA)
-#Copyright (C) 2012-2018 Rui Batista, NV Access Limited, Noelia Ruiz Martínez, Joseph Lee, Babbage B.V.
+#Copyright (C) 2012-2019 Rui Batista, NV Access Limited, Noelia Ruiz Martínez, Joseph Lee, Babbage B.V.
```
Comment (feerrenrut) [inline source/addonHandler/__init__.py]: Please add your name to the copyright list.

File: source/addonHandler/__init__.py [hunk 2]
```diff
@@ -401,7 +402,7 @@ def isRunning(self):
 
 	@property
 	def isDisabled(self):
-		return self.name in _disabledAddons
+		return globalVars.appArgs.disableAddons or self.name in _disabledAddons
```
Comment (feerrenrut) [inline source/addonHandler/__init__.py]: Rather than modify `isDisabled`, which reflects the user preference state, add a condition to `isRunning`

**Comment by DataTriny** [comment]
@DrSooom: yes you can still uninstall add-ons with this PR.

Currently the button to get add-ons and install them is enabled, but I guess it shouldn't, nor the installation process via Windows Explorer. I also think it deserves another PR.

File: source/gui/addonGui.py [hunk 3]
```diff
@@ -296,16 +296,16 @@ def getAddonStatus(self, addon):
 			return incompatibleStatus
 
 		statusList = []
-		if addon.isRunning:
-			# Translators: The status shown for an addon when its currently running in NVDA.
-			statusList.append(_("Enabled"))
-		elif addon.isPendingInstall:
```
Comment (feerrenrut) [inline source/gui/addonGui.py]: Please don't re-arrange this.

**Review by feerrenrut — CHANGES_REQUESTED** [review]
This is a good opportunity to consider if "globally disabled" addons is clearly presented.
Ideally we want to communicate:
- during this run of NVDA, all add-ons are disabled
- what the status of the add-ons will be after a restart, since we expect that to change.

There is a use-case to disable addons, even though `--disable-addons` is present:
When trying to diagnose a broken addon, a user can disable all addons with the command line, then selectively choose a likely stable set / broken candidate to restart to test with.

I think it would be better to explain that all add-ons are disabled,  set the status message to "enabled after restart", "disabled after restart", and continue to allow disabling or enabling the addons.

**Comment by Brian1Gaff** [comment]
OK surely nothing to stop you loading new add ons but they would not be 
enabled orenableable    , if that is a word. Also if an add on is stuck 
enabled are we totally sure that this  code actually does stop that add on 
from even being accessed at all? For the debugging of core vs add ons its 
important this actually works flawlessly I would imagine.
 Brian

bglists@blueyonder.co.uk
Sent via blueyonder.
Please address personal E-mail to:-
briang1@blueyonder.co.uk, putting 'Brian Gaff'
in the display name field.
-----

**Comment by DataTriny** [comment]
@feerrenrut I think I implemented the behavior you want. Note that I intentionally display Disabled state for all add-ons, let me know if you find it confusing with the second state. We could just display this one when `disableAddons` is on afterall.

File: source/addonHandler/__init__.py [hunk 4]
```diff
@@ -388,16 +389,16 @@ def enable(self, shouldEnable):
...
+			# No need to disable an addon that is already disabled, except if all add-ons are disabled.
 			# This also prevents the status in the add-ons dialog from saying "disabled, pending disable"
-			elif self.name not in state["disabledAddons"]:
+			elif self.name not in state["disabledAddons"] or globalVars.appArgs.disableAddons:
 				state["pendingDisableSet"].add(self.name)
 		# Record enable/disable flags as a way of preparing for disaster such as sudden NVDA crash.
 		saveState()
 
 	@property
 	def isRunning(self):
```
Comment (leonardder) [inline source/addonHandler/__init__.py]: Rather than changing this property, could you change isDisabled otherwise? Or will this lead to issues I'm not aware of?
What it will do, is log every add-on that is disabled. I consider this more of a feature than a bug, though.

File: source/gui/addonGui.py [hunk 5]
```diff
@@ -303,16 +303,18 @@ def getAddonStatus(self, addon):
 			# Translators: The status shown for a newly installed addon before NVDA is restarted.
 			statusList.append(_("Install"))
 		# in some cases an addon can be expected to be disabled after install, so we want "install" to take precedence here
-		elif globalVars.appArgs.disableAddons or addon.isDisabled:
+		elif addon.isDisabled or globalVars.appArgs.disableAddons:
```
Comment (leonardder) [inline source/gui/addonGui.py]: Why did you make this change? note that if you touched the disabled property as suggested above, you can change this one as well.

File: source/gui/addonGui.py [hunk 6]
```diff
@@ -303,16 +303,18 @@ def getAddonStatus(self, addon):
...
-		elif globalVars.appArgs.disableAddons or addon.isDisabled:
+		elif addon.isDisabled or globalVars.appArgs.disableAddons:
 			# Translators: The status shown for an addon when its currently suspended do to addons being disabled.
 			statusList.append(_("Disabled"))
 		if addon.isPendingRemove:
 			# Translators: The status shown for an addon that has been marked as removed, before NVDA has been restarted.
 			statusList.append(_("Removed after restart"))
-		elif addon.isPendingDisable or (not addon.isPendingEnable and addon.isPendingInstall and addon.isDisabled):
+		elif addon.isPendingDisable or (addon.isDisabled and not addon.isPendingEnable
+			and (addon.isPendingInstall or globalVars.appArgs.disableAddons)):
```
Comment (leonardder) [inline source/gui/addonGui.py]: Why this change?

File: source/gui/addonGui.py [hunk 7]
```diff
@@ -303,16 +303,18 @@ def getAddonStatus(self, addon):
...
 			# Translators: The status shown for an addon when it requires a restart to become disabled
 			statusList.append(_("Disabled after restart"))
-		elif addon.isPendingEnable or (addon.isPendingInstall and not addon.isDisabled):
+		elif addon.isPendingEnable or (not addon.isDisabled
+			and (addon.isPendingInstall or globalVars.appArgs.disableAddons)):
```
Comment (leonardder) [inline source/gui/addonGui.py]: Why this change?

File: source/gui/addonGui.py [hunk 8]
```diff
@@ -378,7 +380,8 @@ def onClose(self,evt):
 		for addon in self.curAddons:
 			if (addon.isPendingInstall or addon.isPendingRemove
 				or addon.isDisabled and addon.isPendingEnable
-				or addon.isRunning and addon.isPendingDisable):
+				or addon.isRunning and addon.isPendingDisable
+				or (not addon.isRunning and addon.isPendingDisable and globalVars.appArgs.disableAddons)):
```
Comment (leonardder) [inline source/gui/addonGui.py]: Why this change?

**Review by LeonarddeR — DISMISSED** [review]
Thanks for your contribution here!

File: source/addonHandler/__init__.py (hunk 4 above)
Comment (DataTriny) [reply source/addonHandler/__init__.py]: That's what I initially did, but @feerrenrut told me to modify `isRunning` instead.

File: source/gui/addonGui.py (hunk 5 above)
Comment (DataTriny) [reply source/gui/addonGui.py]: Same answer as before: `isDisabled` is meant to reflect the user setting. But I added a comment, to be clearer.

File: source/gui/addonGui.py (hunk 6 above)
Comment (DataTriny) [reply source/gui/addonGui.py]: I wrote this so that add-ons that were previously disabled will have their status indicating they will be disabled after restart (when add-ons are globally disabled).

File: source/gui/addonGui.py (hunk 7 above)
Comment (DataTriny) [reply source/gui/addonGui.py]: Exact same reason as above, I'm not sure if I'm clear because it sounds tricky to explain, but we want to communicate that all add-ons are disabled during this run, but we also want to display if they will be enabled or disabled when you will restart NVDA without the `--disable-addons` flag.

File: source/gui/addonGui.py (hunk 8 above)
Comment (DataTriny) [reply source/gui/addonGui.py]: Because if an add-on was enabled, and you decide to disable it (while add-ons are globally disabled), we want to show the restart dialog. Sorry again, this looks confusing, and I think it is at some point.

**Comment by josephsl** [comment]
Hi,

I'll do a formal review (as the original architect behind disabling add-ons feature).

For reference, for a brief history on how we got here I the first place, please read #3090 discussion and the subsequent pull request I wrote to bring this to life.

Thanks.

**Comment by josephsl** [comment]
Hi,

It turns out this is a major oversight from #6275 - a regression, in fact. Just because an add-on is deemed incompatible by consulting compatibility flags (or lack thereof) should not be a reason to ignore --disable-addons command line switch. Regression confirmed with NVDA 2018.4 - in 2018.4, with --disable-addons switch on, all add-ons are listed as "disabled".

To @DataTriny: when filing an issue, please try testing the issue with earlier NVDA releases. Doing so will change your outlook on pull requests like this and might help all of us understand the big picture better.

Thanks.

**Comment by josephsl** [comment]
Hi,

To @LeonardDer: after looking at the code and comments closely, I can see why Reef suggested modifying isRunning property, although I do agree that isDisabled would have been a better place to do so (in the interest of reducing code duplication).

Here's what the "add-ons disabler" algorithm used to do (2018.4), current situation (2019.1.1), and what this pull request proposes:

2018.4:

* Add-on handler:
 * isRunning: as long as the add-on is about to be installed after a restart or not disabled
 * isDisabled: add-on name is not recorded in disabled add-ons set
* Add-on GUI:
 * Add-on status: all add-ons are enabled (i.e. isRunning is assumed to be the default case). The "disabled" flag is returned if --disable-addons switch is present or add-on reports itself as disabled (note the "or" operator, which will be evaluated left to right in Python; that is, --disable-addons switch overrides individual add-ons, thus works as expected as specified in #3090).

2019.1.1:

* Add-on handler:
 * isRunning: compatibility block is checked in addition to above.
 * isDisabled: same as before. The very fact that --disable-addons switch is not checked is the actual root of this regression.
* Add-on GUI:
 * Add-on status: "isBlocked" takes precedence. Due to introduction of restart action flags, status list is split in two, with isRunning/isDisabled flags consulted first before restart action is added.

Pull request:

* Add-on handler:
 * isRunning: in addition to 2019.1.1, --disable-addons switch will be checked.
 * isDisabled: no change from 2018.4
* Add-on GUI:
 * Add-on status: when computing pending enable/disable flags, --disable-addons flag will also be checked.

I think there is a slightly simpler short-term solution, and a long-term solution that requires a fundamental change as to how we view add-ons that are actually running:

Short-term: just modify the status text retriever. This can be done as follows:

1. Before creating status list, have a boolean that checks addon.isRunning and --disable-addons switch is present.
2. When looking up secondary status texts, for "enable/disable after restart", check the just created boolean flag and unconditionally announce this flag if the add-on was previously enabled/disabled in normal operation.

This is perhaps the easiest (and a naive) solution that resolves this regression, with the obvious downside being deceiving add-on developers with confusing add-on active flags (a potential ethical problem down the road).

Long-term solution: split addon.isRunning property into two parts:

1. addon.isReady: rename isRunning property to isReady property.
2. isRunning: addon.isReady and not globalVars.appArgs.disableAddons.

This solution is clearer at the cost of introducing incompatible changes. NVDA Core routines and add-ons that rely on current behavior of addon.isRunning must be changed to obtain addon.isReady (this affects not only this pull request, but also #3208).

Thanks.

**Comment by DataTriny** [comment]
Hi everyone,

@josephsl sorry not to have taken the time to test with an older version of NVDA. I'll try to always test with the previous version. I confirm this is a regression, however you can't neither disable an add-on nor see its state after restart.

I agree that modifying `isRunning` forces some code duplication that may introduce bugs in the future. I would rather directly implement your second solution which introduce `isReady` or `canRun`. I heard that transition to Python 3 is about to begin, and of course all add-ons will have to be ported anyway, so it could be the perfect occasion to introduce this breaking change.

File: source/addonHandler/__init__.py (hunk 4 above)
Comment (feerrenrut) [reply source/addonHandler/__init__.py]: Is disabled relates to the users saved state. `isRunning` does not, instead `isRunning` is the negation of several other reasons it might not be running: eg blocked, disabled, pending install

Comment (leonardder) [reply source/addonHandler/__init__.py]: Ah, that's fine then.

**Comment by feerrenrut** [comment]
> It turns out this is a major oversight from #6275 - a regression, in fact. Just because an add-on is deemed incompatible by consulting compatibility flags (or lack thereof) should not be a reason to ignore --disable-addons command line switch.

Lets not overreact here. Testing this locally shows that the addons are not running. The confusion is caused by trying to use one word to communicate two concepts:
- The user set state: enabled / disabled
- The global override of the user set state with `--disable-addons`

They cant run because of the guard in `def addConfigDirsToPythonPackagePath(module, subdir=None)` of `source/config/__init__.py` which if not in place would call through to:
`def addToPackagePath(self, package)` in `addonHandler/__init__.py` which is where the real "enabling" happens.

This regression is just a change in what is deemed more important to show and was intentionally, though not well documented or highlighted for community discussion. After-all there was a lot of other parts to focus on.

I would prefer that you make only the minimal changes required to clarify the difference between globally disabled and user state:

When `--disable-addons` is present:
 - Add a static text before the list view: "NVDA was started with all add-ons disabled, no add-ons will run until NVDA is restarted. You may modify the enabled / disabled state, install or uninstall add-ons, which will take effect after NVDA is restarted."
- Change the text for "enabled" to "enabled after restart"
- Change the text for "disabled" to "disabled after restart"

I believe this will clear up the confusion.

**Comment by feerrenrut** [comment]
I just realised, we already have the "all add-ons disabled message". It currently says: "All add-ons are currently disabled. To enable add-ons you must restart NVDA." This is pretty clear to me.

Do we want to add the other bit about changing state: " You may modify the enabled / disabled state, install or uninstall add-ons, which will take effect after NVDA is restarted."?

**Comment by josephsl** [comment]
Hi, yep, apologies for possible bad reaction. How about clarifying this in the window title (say, “Add-ons Manager – all add-ons disabled”)? Although a bit verbose, this might communicate the fact that the command line switch is in effect. This is sort of akin to desktop watermark shown when Windows starts in safe mode. Thanks.

**Comment by josephsl** [comment]
Hi, if this makes things clearer for a beginner, yes, along with documenting this fact in the user guide. Thanks,

**Comment by feerrenrut** [comment]
>  How about clarifying this in the window title

Yes, I like that idea as well.

**Comment by lukaszgo1** [comment]
With the latest commit #5785 is also fixed isn't it?

**Comment by DataTriny** [comment]
@feerrenrut I removed the "Disabled" state when add-ons are globally disabled, I modified the static text that appears on top of the list view (the one you suggested), and I conditionally changed the title of the window. I am not 100% confident however with the text you suggested because it feels a bit weird to me, with "NVDA" and "restart" repeated multiple times, but I am not a native English speaker, so I will let you decide. Since I modified two translatable strings, should I do something to warn translators or whatever?

**Comment by josephsl** [comment]
Hi, just add a translator comment for new ones, and if needed, change the translator comment associated with changed messages. Thanks.

**Comment by josephsl** [comment]
Hi,

Yes, this resolves #5785, but then it broke Appveyor build for some reason (investigating).

Thanks.

**Comment by DataTriny** [comment]
The only time I didn't run tests, it fails. Should pass now!

File: source/gui/addonGui.py [hunk 9]
```diff
@@ -378,7 +383,8 @@ def onClose(self,evt):
 		for addon in self.curAddons:
 			if (addon.isPendingInstall or addon.isPendingRemove
 				or addon.isDisabled and addon.isPendingEnable
-				or addon.isRunning and addon.isPendingDisable):
+				or addon.isRunning and addon.isPendingDisable
```
Comment (feerrenrut) [inline source/gui/addonGui.py]: I think you could leave out the appArgs line and just change this line:
```suggestion
				or not addon.isDisabled and addon.isPendingDisable
```

File: source/gui/addonGui.py [hunk 10]
```diff
@@ -157,15 +157,17 @@ def __init__(self, parent):
 		AddonsDialog._instance = weakref.ref(self)
 		# Translators: The title of the Addons Dialog
 		title = _("Add-ons Manager")
-		wx.Dialog.__init__(self, parent, title=title)
+		# Translators: The title of the Addons Dialog when add-ons are globally disabled
+		titleWhenAddonsAreDisabled = _("Add-ons Manager (add-ons globally disabled)")
```
Comment (feerrenrut) [inline source/gui/addonGui.py]: I think `globally` is a little to dev-centric. I think "all add-ons disabled" or maybe just "add-ons disabled"

File: source/gui/addonGui.py [hunk 11]
```diff
@@ -157,15 +157,17 @@ def __init__(self, parent):
...
 		DpiScalingHelperMixin.__init__(self, self.GetHandle())
 
 		mainSizer = wx.BoxSizer(wx.VERTICAL)
 		firstTextSizer = wx.BoxSizer(wx.VERTICAL)
 		listAndButtonsSizerHelper = guiHelper.BoxSizerHelper(self, sizer=wx.BoxSizer(wx.HORIZONTAL))
 		if globalVars.appArgs.disableAddons:
-			# Translators: A message in the add-ons manager shown when all add-ons are disabled.
-			label = _("All add-ons are currently disabled. To enable add-ons you must restart NVDA.")
+			# Translators: A message in the add-ons manager shown when add-ons are globally disabled.
+			label = _("NVDA was started with all add-ons disabled, no add-ons will run until NVDA is restarted. You may modify the enabled / disabled state, install or uninstall add-ons, which will take effect after NVDA is restarted.")
```
Comment (feerrenrut) [inline source/gui/addonGui.py]: I agree with your other comment:

>it feels a bit weird to me, with "NVDA" and "restart" repeated multiple times

Perhaps something like:
```
NVDA was started with all add-ons disabled. You may modify the enabled / disabled state, and install or uninstall add-ons. Changes will not take effect after NVDA is restarted.
```

File: source/gui/addonGui.py [hunk 12]
```diff
@@ -157,15 +157,17 @@ def __init__(self, parent):
 		AddonsDialog._instance = weakref.ref(self)
 		# Translators: The title of the Addons Dialog
 		title = _("Add-ons Manager")
-		wx.Dialog.__init__(self, parent, title=title)
+		# Translators: The title of the Addons Dialog when add-ons are globally disabled
+		titleWhenAddonsAreDisabled = _("Add-ons Manager (add-ons globally disabled)")
+		wx.Dialog.__init__(self, parent, title=title if not globalVars.appArgs.disableAddons else titleWhenAddonsAreDisabled)
```
Comment (feerrenrut) [inline source/gui/addonGui.py]: Can you split this across multiple lines please?

```suggestion
		wx.Dialog.__init__(
			self,
			parent,
			title=title if not globalVars.appArgs.disableAddons else titleWhenAddonsAreDisabled
		)
```

File: source/gui/addonGui.py [hunk 13]
```diff
@@ -303,16 +305,19 @@ def getAddonStatus(self, addon):
 			# Translators: The status shown for a newly installed addon before NVDA is restarted.
 			statusList.append(_("Install"))
 		# in some cases an addon can be expected to be disabled after install, so we want "install" to take precedence here
-		elif globalVars.appArgs.disableAddons or addon.isDisabled:
+		# If add-ons are globally disabled, don't show this status.
+		elif addon.isDisabled and not globalVars.appArgs.disableAddons:
```
Comment (feerrenrut) [inline source/gui/addonGui.py]: Because you moved the `appArgs.disableAddons` condition out of `isDisabled`, we no longer have to guard against it here:
```suggestion
		elif addon.isDisabled:
```
Also remove the comment above.

**Comment by feerrenrut** [comment]
If you like I could make these changes for you? There are a handful of comments I would like to add to this code as well.

File: source/gui/addonGui.py (hunk 13 above)
Comment (DataTriny) [reply source/gui/addonGui.py]: No we can't remove this guard or otherwise, if an add-on is disabled we will display "Disabled, Disabled after restart".

Comment (feerrenrut) [reply source/gui/addonGui.py]: ah, yes of course.

**Comment by DataTriny** [comment]
I made all the changes you requested, but if you still want to add something...

**Comment by josephsl** [comment]
Hi, I can also assist in editing the user guide and user interface messages if you want. Thanks.

**Comment by DataTriny** [comment]
@josephsl UI messages should be fine now. Looking through the user guide, I don't see what I could add as the `--disable-addons` is mentioned, although briefly.

File: source/gui/addonGui.py [hunk 14]
```diff
@@ -157,15 +157,21 @@ def __init__(self, parent):
...
 		DpiScalingHelperMixin.__init__(self, self.GetHandle())
 
 		mainSizer = wx.BoxSizer(wx.VERTICAL)
 		firstTextSizer = wx.BoxSizer(wx.VERTICAL)
 		listAndButtonsSizerHelper = guiHelper.BoxSizerHelper(self, sizer=wx.BoxSizer(wx.HORIZONTAL))
 		if globalVars.appArgs.disableAddons:
-			# Translators: A message in the add-ons manager shown when all add-ons are disabled.
-			label = _("All add-ons are currently disabled. To enable add-ons you must restart NVDA.")
+			# Translators: A message in the add-ons manager shown when add-ons are globally disabled.
+			label = _("NVDA was started with all add-ons disabled. You may modify the enabled / disabled state, and install or uninstall add-ons. Changes will not take effect after NVDA is restarted.")
```
Comment (feerrenrut) [inline source/gui/addonGui.py]: oops, it looks like I skipped an important word here, "until". Try the following:
```suggestion
			label = _("NVDA was started with all add-ons disabled. You may modify the enabled / disabled state, and install or uninstall add-ons. Changes will not take effect until after NVDA is restarted.")
```

Comment (feerrenrut) [reply source/gui/addonGui.py]: It's actually long enough that it would be nice to have it split up. I would split it at the sentence boundaries:

```suggestion
			label = _(
				"NVDA was started with all add-ons disabled. "
				"You may modify the enabled / disabled state, and install or uninstall add-ons. "
				"Changes will not take effect until after NVDA is restarted."
			)
```

File: source/gui/addonGui.py [hunk 15]
```diff
@@ -303,16 +309,19 @@ def getAddonStatus(self, addon):
...
-		elif globalVars.appArgs.disableAddons or addon.isDisabled:
+		# If add-ons are globally disabled, don't show this status.
+		elif addon.isDisabled and not globalVars.appArgs.disableAddons:
 			# Translators: The status shown for an addon when its currently suspended do to addons being disabled.
 			statusList.append(_("Disabled"))
 		if addon.isPendingRemove:
 			# Translators: The status shown for an addon that has been marked as removed, before NVDA has been restarted.
 			statusList.append(_("Removed after restart"))
-		elif addon.isPendingDisable or (not addon.isPendingEnable and addon.isPendingInstall and addon.isDisabled):
+		elif addon.isPendingDisable or (addon.isDisabled and not addon.isPendingEnable
```
Comment (feerrenrut) [inline source/gui/addonGui.py]: I would probably write this in a slightly longer way, to improve the clarity:
```suggestion
		elif addon.isPendingEnable or (
				# yet to be installed, enabled after install
				addon.isPendingInstall and not addon.isDisabled
			) or (
				# addons globally disabled, enabled after restart
				globalVars.appArgs.disableAddons and not addon.isDisabled
			):
```

Comment (feerrenrut) [reply source/gui/addonGui.py]: Note, the first time I wrote on the wrong line. I have updated the comment.

File: source/gui/addonGui.py [hunk 16]
```diff
@@ -303,16 +309,19 @@ def getAddonStatus(self, addon):
...
+			and (addon.isPendingInstall or globalVars.appArgs.disableAddons)):
 			# Translators: The status shown for an addon when it requires a restart to become disabled
 			statusList.append(_("Disabled after restart"))
-		elif addon.isPendingEnable or (addon.isPendingInstall and not addon.isDisabled):
+		elif addon.isPendingEnable or (not addon.isDisabled
```
Comment (feerrenrut) [inline source/gui/addonGui.py]: Also here, slightly longer for clarity, with comments:
```suggestion
		elif addon.isPendingEnable or (
				# yet to be installed, enabled after install
				addon.isPendingInstall and not addon.isDisabled
			) or (
				# addons globally disabled, enabled after restart
				globalVars.appArgs.disableAddons and not addon.isDisabled
			):
```

**Review by feerrenrut — CHANGES_REQUESTED** [review]
I think that's everything now. I expect this has been a slightly frustrating review for you, sorry for that. But thank you again for your contributions!

File: source/gui/addonGui.py (hunk 14 above)
Comment (DataTriny) [reply source/gui/addonGui.py]: My bad, I didn't even read your suggestion, just copy and paste...

File: source/gui/addonGui.py (hunk 15 above)
Comment (DataTriny) [reply source/gui/addonGui.py]: Nice try, but you just missed a condition, otherwise we would always display "Disabled after restart" when add-on is disabled in the settings.

**Comment by DataTriny** [comment]
I really appreciate your menthoring, don't worry. I spend a lot of time observing before doing, but your suggestions point me in the right direction! Thank you!

The longest text on the top makes me see that we now have room on the right, and that columns on the list view are too small to display even the version. We might consider expanding the width of the list view.

**Comment by feerrenrut** [comment]
I pushed a change to wrap that text. I think there are a few other modifications to the gui that could be made. Though I think they should be in another PR:
- Adjusting the width and making the dialog resizable.
- Experimenting with removing columns, and putting information from the dialog which launches after the "about addon" button is pressed into another GUI element underneath.

**Comment by feerrenrut** [comment]
Let me know if you are happy with this and I will do a final review before merge.

**Comment by DataTriny** [comment]
Yes I'm OK with the changes. I didn't mean to make any UI design on this PR, but I share your ideas on it.

File: source/addonHandler/__init__.py [hunk 17]
```diff
@@ -388,7 +389,7 @@ def enable(self, shouldEnable):
 			if self.name in state["pendingEnableSet"]:
 				# Undoing a pending enable.
 				state["pendingEnableSet"].discard(self.name)
-			# No need to disable an addon that is already disabled.
+			# No need to disable an addon that is already disabled, except if all add-ons are disabled.
```
Comment (feerrenrut) [inline source/addonHandler/__init__.py]: This comment is not necessary now.
```suggestion
			# No need to disable an addon that is already disabled.
```

**Review by feerrenrut — APPROVED** [review]
